### PR TITLE
implements session object handle iterator, with caching

### PR DESCRIPTION
**Pull request by keldonin**
Hello, 

This PR is an alternate candidate for implementing #106.

It implements the Iterator trait for object handlers, given a search template. The iterator is able to fetch values ahead, and cache the results. The cache size can be specified at iterator build time.

## design choices

- The iterator will return an `Option<Result<ObjectHandle>>`. This is because calls to `C_FindObjects()` may fail, and the caller may want to manage such condition. It is similar to what the  [`std::io::Lines`](https://doc.rust-lang.org/std/io/struct.Lines.html#method.next) iterator implements.
- I didn't replace the `find_objects()` method. It seems to me it serves a different goal,  ease of use. On the other hand, the iterator allows the caller a more fine-grained control over how object handles are retrieved from the token, and allows for streaming. Both can coexist, IMO.
- as with the recent `find_objects` change, every effort is made to avoid unnecessary calls to `C_FindObjects()`. 
- The iterator implements the `Drop` trait, to call `C_FindObjectsFinal()`.
- Thanks to the design choices made for the Session object, the iterator features no locking mechanism for its state variables. It assumes that the iterator object will never be accessed concurrently from different threads.
-  No code was added to prevent creating two iterators: It relies upon the underlying token to prevent running two searches simultaneously.
- Because the `next()` method of an iterator cannot return a `Result`, I had to create a new macro from `get_pkcs11!()` that does not carry the syntactic sugar `?`.
- The implementation tries to address the unlikely case when `C_FindObjects` and `C_FindObjectsFinal` are not implemented. But... perhaps the library should refuse to work with such library, and detect that at library loading time? 

For your review,

### CODE REVIEW

File: cryptoki/tests/basic.rs
Comment (wiktor-k) [inline cryptoki/tests/basic.rs:371]:
```suggestion
fn session_objecthandle_iterator() -> testresult::TestResult {
```

(and `Ok(())` at the end)

And then (almost) all `unwrap`s can just be converted to `?`.

File: cryptoki/src/session/object_management.rs
Comment (wiktor-k) [inline cryptoki/src/session/object_management.rs:223]:
```suggestion
        let template: Vec<CK_ATTRIBUTE> = template.iter().map(Into::into).collect();
```

I think this would be shorter :thinking:

**Review by wiktor-k — APPROVED** [review]
Okay, I think it looks very nice :+1: 

I've added a couple of personal, tiny nitpicks.

I also wonder if it'd be a good idea to add `Closes: https://github.com/parallaxsecond/rust-cryptoki/pull/109` in the commit message trailer (since it appears to be the same functionality but with a separate method).

One further improvement would be re-implementing `find_objects` in terms of your `iter_objects` (it should be as simple as just calling `iter_objects` and then `collect` on the result :thinking: )

Thanks for your time! :wave:

**Comment by wiktor-k** [comment]
> No code was added to prevent creating two iterators: It relies upon the underlying token to prevent running two searches simultaneously.

Correct me if I'm wrong but this should be as simple as making `iter_objects_with_cache_size` take a `mut` ref to Session. This way as long as the `ObjectHandleIterator` exists (which would hold a `mut` ref to Session no other iterators could be created (since `mut` means "exclusive"). :thinking:

**Comment by keldonin** [comment]
> > No code was added to prevent creating two iterators: It relies upon the underlying token to prevent running two searches simultaneously.
> 
> Correct me if I'm wrong but this should be as simple as making `iter_objects_with_cache_size` take a `mut` ref to Session. This way as long as the `ObjectHandleIterator` exists (which would hold a `mut` ref to Session no other iterators could be created (since `mut` means "exclusive"). 🤔

That would work. However, if I combine this with a rewrite of `find_objects()` as suggested above, it will propagate the need for a mutable ref, leading to a breaking change for that API: 

```
pub fn find_objects(&mut self, template: &[Attribute]) -> Result<Vec<ObjectHandle>>`
```

Choose your curse 😉

**Comment by keldonin** [comment]
Actually there is maybe a good reason for not taking a mutable ref:

```
    for obj in session.iter_objects(&token_object)? {
        let obj = obj?; 

        let attributes = session.get_attributes(obj, &wanted_attr)?;
        ...
```

In this pattern, attributes are retrieved as objects are fetched. It results in interleaved calls to `C_FindObjects()` and `C_GetAttributeValue()`, which is legit. In fact, crypto operations are also legit during a search. implementing session with a mutable ref would prevent this, ans session.get_attributes() is also needing a ref to the session.


If preventing concurrent searches by the library itself is desirable, it can be easily implemented using a flag in session, inside a `Cell` for interior mutability. 

Any thought?

**Comment by wiktor-k** [comment]
Yeah, all good points. Maybe we could just leave the "no multiple open searches" invariant in docs and that's it. Making it too complex to cover a rare API misuse may not be such a great idea after all :thinking: 

(I've got a couple of complex ideas how to "solve" it but I worry that the cure will be worse than the disease :sweat_smile: )

File: cryptoki/src/session/object_management.rs
Comment (keldonin) [reply cryptoki/src/session/object_management.rs:33]: I'm sorry, but I'm unclear on this one. What is the change about? and what about the trailing `main()` at the end of the example?

File: cryptoki/src/session/object_management.rs
Comment (keldonin) [reply cryptoki/src/session/object_management.rs:87]: Not sure, TBH. I think this will grow verbosity. I couldn't spot any obvious benefit, unless I missed something?

File: cryptoki/tests/basic.rs
Comment (keldonin) [reply cryptoki/tests/basic.rs:371]: That's a good one!

File: cryptoki/src/session/object_management.rs
Comment (keldonin) [reply cryptoki/src/session/object_management.rs:33]: OK I get it now.

**Comment by keldonin** [comment]
> I also wonder if it'd be a good idea to add `Closes: https://github.com/parallaxsecond/rust-cryptoki/pull/109` in the commit message trailer (since it appears to be the same functionality but with a separate method).

When all the reviews are finished, I can rebase the branch and incorporate the comment.

> One further improvement would be re-implementing `find_objects` in terms of your `iter_objects` (it should be as simple as just calling `iter_objects` and then `collect` on the result 🤔 )
> 

That's been incorporated.

File: cryptoki/src/session/object_management.rs
Comment (keldonin) [reply cryptoki/src/session/object_management.rs:89]: Will push a commit for this.

File: cryptoki/src/session/object_management.rs
Comment (ionut-arm) [inline cryptoki/src/session/object_management.rs:176]: I'm not sure why this log is here, isn't this branch the one where `C_FindObjectsFinal` exists?

**Review by ionut-arm — COMMENTED** [review]
Approach looks good to me, happy to go ahead with it, just a few minor comments

File: cryptoki/src/session/object_management.rs
Comment (wiktor-k) [inline cryptoki/src/session/object_management.rs:178]:
```suggestion
            if let Rv::Error(error) = Rv::from(unsafe { f(self.session.handle()) }) {
                log::error!(...);
            }
```

I meant something like this, sorry I wasn't specific :)

File: cryptoki/src/session/object_management.rs
Comment (wiktor-k) [inline cryptoki/src/session/object_management.rs:248]: I wonder if this would suffice:

```suggestion
            .collect()
```

(since it's in a return position and the type of the result and this is identical)

File: cryptoki/src/session/object_management.rs
Comment (wiktor-k) [inline cryptoki/src/session/object_management.rs:15]: AFAIK `NonZeroUsize::new` is const and could be directly initialized here:

```suggestion
const MAX_OBJECT_COUNT: NonZeroUsize = NonZeroUsize::new(10).expect("non-zero value to be passed");
```

**Review by wiktor-k — COMMENTED** [review]
I think it's the last round of feedback and in general it looks very nice already :ok_hand:

File: cryptoki/src/session/object_management.rs
Comment (keldonin) [reply cryptoki/src/session/object_management.rs:25]: Sure I can add it. However the `new()` method is not public, it will not appear in the generated doc.

File: cryptoki/src/session/object_management.rs
Comment (keldonin) [reply cryptoki/src/session/object_management.rs:178]: good catch.

File: cryptoki/src/session/object_management.rs
Comment (keldonin) [reply cryptoki/src/session/object_management.rs:15]: `expect()` is not stable yet to be assigned to a `const`. In this instance, I believe it is OK to use `new_unchecked()` and document the safety condition in the code.

File: cryptoki/src/session/object_management.rs
Comment (keldonin) [reply cryptoki/src/session/object_management.rs:176]: resolved - see other related comment.

**Review by wiktor-k — APPROVED** [review]
LGTM :+1: thanks for taking the time to address all remarks :)

**Comment by keldonin** [comment]
@wiktor-k: 

NP, feedback is useful and relevant!

I don't know what is your policy for merging changes, but since you made this request:

> I also wonder if it'd be a good idea to add Closes: https://github.com/parallaxsecond/rust-cryptoki/pull/109 in the commit message trailer (since it appears to be the same functionality but with a separate method).

I suggested I would rebase this PR, but that would make another round of approvals, and you would loose visibility on the changes before merging. Maybe is it simpler for you to squash merge and add this comment to the commit message?

File: cryptoki/src/context/mod.rs
Comment (hug-dev) [inline cryptoki/src/context/mod.rs:18]: Would it work to modify `get_pkcs11` macro to use this one + the `ok_or`? Meaning to avoid code duplication :)

File: cryptoki/tests/basic.rs
Comment (hug-dev) [inline cryptoki/tests/basic.rs:390]: I noticed that you removed the `unwrap` here, is it ok?

File: cryptoki/src/session/object_management.rs
Comment (hug-dev) [inline cryptoki/src/session/object_management.rs:161]: Since this is done in both branches, maybe this line can be done before and so we just need the `if` condition with the `break`?

File: cryptoki/src/session/object_management.rs
Comment (hug-dev) [inline cryptoki/src/session/object_management.rs:309]: Nice to re-use what you did here!

**Review by hug-dev — APPROVED** [review]
Nitpicking comments you can feel free to ignore if they are not worth it! Otherwise trusting Wiktor's review :pray:

File: cryptoki/src/session/object_management.rs
Comment (wiktor-k) [reply cryptoki/src/session/object_management.rs:161]: Haha I had the same thought! But then the if condition depends on the variable that'd be overwritten (so the old value would've to be stored in a temporary variable)... :)

File: cryptoki/tests/basic.rs
Comment (wiktor-k) [reply cryptoki/tests/basic.rs:390]: Yeah, let's use `?` here. The warning is suppressed because of the variable assignment but better to check IMO.

**Comment by wiktor-k** [comment]
> Otherwise trusting Wiktor's review 🙏

Haha, please don't do that 😅 I try to be diligent here but it's still an unpaid side-project so my time is limited.

I still think that given our collective effort the code continously gets better and better (and with tests future refactoring is always possible).

👋

File: cryptoki/src/session/object_management.rs
Comment (keldonin) [reply cryptoki/src/session/object_management.rs:161]: I couldn't give a better answer 😅

File: cryptoki/src/session/object_management.rs
Comment (keldonin) [reply cryptoki/src/session/object_management.rs:309]: Rewriting `find_objects` as collection of the iterator was an idea suggested by @wiktor-k. And yes, the code here was part of a previous PR, right? I guess some of it belongs to the iterator code now.

File: cryptoki/src/context/mod.rs
Comment (keldonin) [reply cryptoki/src/context/mod.rs:18]: Seems reasonable, I will give it a stab. Not much XP with Rust macro authoring, however. I'll have a look at it this week-end.

File: cryptoki/tests/basic.rs
Comment (keldonin) [reply cryptoki/tests/basic.rs:390]: @hug-dev good catch. I will need to refactor a bit, as the closure don't like handling `?`.

File: cryptoki/src/context/mod.rs
Comment (keldonin) [reply cryptoki/src/context/mod.rs:18]: here is it - easier than anticipated it seems.

**Review by wiktor-k — APPROVED** [review]

File: cryptoki/src/session/object_management.rs
Comment (hug-dev) [reply cryptoki/src/session/object_management.rs:161]: ahh got it!! Programming is not as easy as it seems :eyeglasses:

**Review by hug-dev — APPROVED** [review]

**Comment by keldonin** [comment]
Thank you for merging this one!

Out of curiosity, are you planning to make a new tagged release at some point in time?

**Comment by hug-dev** [comment]
Nothing planned for now I think, it's mostly demand-based! We also try to group release with other dependencies in `parallaxsecond`. If you need one for your projects, we can start to see when will work!

**Comment by hug-dev** [comment]
On another topic @keldonin, would you like to be a maintainer of this crate? Since you have shown interest in it in the past few weeks with very nice improvements I was wondering if you would like to help us maintaining it?
There is no special asks or responsibility apart from the power to be able to approve and merge PRs!

**Comment by keldonin** [comment]
> Nothing planned for now I think, it's mostly demand-based! We also try to group release with other dependencies in `parallaxsecond`. If you need one for your projects, we can start to see when will work!

Hi @hug-dev. Nothing pressing for now, just asking. Thank you for the clarification!

**Comment by keldonin** [comment]
> On another topic @keldonin, would you like to be a maintainer of this crate? Since you have shown interest in it in the past few weeks with very nice improvements I was wondering if you would like to help us maintaining it? There is no special asks or responsibility apart from the power to be able to approve and merge PRs!

I would be honoured.

Time-wise, something I can perform during my free time (probably like you? 🤔) - that I share with other activities - there is life out there 😂. 

So, glad to give support here, to the extent of my availability.

**Comment by hug-dev** [comment]
That's great, thank you!
There is absolutely no ask from our side, put into it as much time as you want! It's mostly so that we can share the merging power so there is less chance that a PR stays too long in limbo state/unreviewed :smiley:

**Comment by hug-dev** [comment]
It's done now! If you want, you can join the `#parsec` Slack channel on the CNCF Slack workspace :) More info [here](https://github.com/parallaxsecond/community?tab=readme-ov-file#community-channel)

**Pull request merged**
